File: attocode/src/providers/adapters/anthropic.ts
/**
 * Anthropic Claude Provider Adapter
 *
 * Adapts the Anthropic SDK to our LLMProvider interface.
 * Supports native tool use via the chatWithTools method.
 */

import type {
  LLMProvider,
  LLMProviderWithTools,
  Message,
  MessageWithContent,
  ChatOptions,
  ChatOptionsWithTools,
  ChatResponse,
  ChatResponseWithTools,
  ToolCallResponse,
  ToolDefinitionSchema,
  AnthropicConfig
} from '../types.js';
import { ProviderError } from '../types.js';
import { registerProvider, hasEnv, requireEnv } from '../provider.js';
import { resilientFetch, type NetworkConfig } from '../resilient-fetch.js';

// =============================================================================
// ANTHROPIC API TYPES
// =============================================================================

/** Anthropic tool definition format */
interface AnthropicTool {
  name: string;
  description: string;
  input_schema: Record<string, unknown>;
}

/** Anthropic content block types */
type AnthropicContentBlock =
  | { type: 'text'; text: string }
  | { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> }
  | { type: 'tool_result'; tool_use_id: string; content: string };

// =============================================================================
// ANTHROPIC PROVIDER
// =============================================================================

export class AnthropicProvider implements LLMProvider, LLMProviderWithTools {
  readonly name = 'anthropic';
  readonly defaultModel = 'claude-sonnet-4-20250514';

  private apiKey: string;
  private model: string;
  private baseUrl: string;
  private networkConfig: NetworkConfig;

  constructor(config?: AnthropicConfig) {
    this.apiKey = config?.apiKey ?? requireEnv('ANTHROPIC_API_KEY');
    this.model = config?.model ?? this.defaultModel;
    this.baseUrl = config?.baseUrl ?? 'https://api.anthropic.com';
    // Anthropic requests can be slow for complex tasks
    this.networkConfig = {
      timeout: 120000,  // 2 minutes for complex reasoning
      maxRetries: 3,
      baseRetryDelay: 1000,
    };
  }

  isConfigured(): boolean {
    return hasEnv('ANTHROPIC_API_KEY');
  }

  async chat(messages: (Message | MessageWithContent)[], options?: ChatOptions): Promise<ChatResponse> {
    const model = options?.model ?? this.model;

    // Separate system message from conversation
    const systemMessage = messages.find(m => m.role === 'system');
    const conversationMessages = messages
      .filter(m => m.role !== 'system')
      .map(m => ({
        role: m.role as 'user' | 'assistant',
        content: typeof m.content === 'string' ? m.content : m.content.map(c => c.text).join(''),
      }));

    // Build system content — supports structured blocks with cache_control for prompt caching
    let systemContent: string | Array<{ type: 'text'; text: string; cache_control?: { type: 'ephemeral' } }> | undefined;
    if (systemMessage) {
      if (typeof systemMessage.content !== 'string' && Array.isArray(systemMessage.content)) {
        // Structured content with cache_control markers
        systemContent = systemMessage.content;
      } else {
        systemContent = typeof systemMessage.content === 'string'
          ? systemMessage.content
          : (systemMessage.content as Array<{ text: string }>).map(c => c.text).join('');
      }
    }

    // Build request body
    const body = {
      model,
      max_tokens: options?.maxTokens ?? 4096,
      temperature: options?.temperature ?? 0.7,
      system: systemContent,
      messages: conversationMessages,
      ...(options?.stopSequences && { stop_sequences: options.stopSequences }),
    };

    try {
      const { response } = await resilientFetch({
        url: `${this.baseUrl}/v1/messages`,
        init: {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': this.apiKey,
            'anthropic-version': '2023-06-01',
            'anthropic-beta': 'prompt-caching-2024-07-31',
          },
          body: JSON.stringify(body),
        },
        providerName: this.name,
        networkConfig: this.networkConfig,
        onRetry: (attempt, delay, error) => {
          console.warn(`[Anthropic] Retry attempt ${attempt} after ${delay}ms: ${error.message}`);
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw this.handleError(response.status, error);
      }

      const data = await response.json() as {
        content: Array<{ type: string; text: string }>;
        stop_reason: string;
        usage: {
          input_tokens: number;
          output_tokens: number;
          cache_creation_input_tokens?: number;
          cache_read_input_tokens?: number;
        };
      };

      // Extract text from content blocks
      const content = data.content
        .filter(block => block.type === 'text')
        .map(block => block.text)
        .join('');

      return {
        content,
        stopReason: this.mapStopReason(data.stop_reason),
        usage: {
          inputTokens: data.usage.input_tokens,
          outputTokens: data.usage.output_tokens,
          cacheReadTokens: data.usage.cache_read_input_tokens,
          cacheWriteTokens: data.usage.cache_creation_input_tokens,
        },
      };
    } catch (error) {
      if (error instanceof ProviderError) throw error;
      throw new ProviderError(
        `Anthropic request failed: ${(error as Error).message}`,
        this.name,
        'NETWORK_ERROR',
        error as Error
      );
    }
  }

  /**
   * Send a chat request with tool definitions.
   * Anthropic's native tool use format differs from OpenAI's:
   * - Tools use `input_schema` instead of `parameters`
   * - Tool calls are in content blocks with type `tool_use`
   * - Tool results are sent as `tool_result` content blocks
   */
  async chatWithTools(
    messages: (Message | MessageWithContent)[],
    options?: ChatOptionsWithTools
  ): Promise<ChatResponseWithTools> {
    const model = options?.model ?? this.model;

    // Separate system message from conversation
    const systemMessage = messages.find(m => m.role === 'system');

    // Convert messages to Anthropic format
    const anthropicMessages = this.convertMessagesToAnthropicFormat(
      messages.filter(m => m.role !== 'system')
    );

    // Convert tool definitions from OpenAI format to Anthropic format
    const anthropicTools = options?.tools?.map(this.convertToolToAnthropicFormat);

    // Build system content — supports structured blocks with cache_control for prompt caching
    let systemContent: string | Array<{ type: 'text'; text: string; cache_control?: { type: 'ephemeral' } }> | undefined;
    if (systemMessage) {
      if (typeof systemMessage.content !== 'string' && Array.isArray(systemMessage.content)) {
        // Structured content with cache_control markers - pass through directly
        systemContent = systemMessage.content;
      } else {
        systemContent = typeof systemMessage.content === 'string'
          ? systemMessage.content
          : (systemMessage.content as Array<{ text: string }>).map(c => c.text).join('');
      }
    }

    // Build request body
    const body: Record<string, unknown> = {
      model,
      max_tokens: options?.maxTokens ?? 4096,
      temperature: options?.temperature ?? 0.7,
      system: systemContent,
      messages: anthropicMessages,
      ...(options?.stopSequences && { stop_sequences: options.stopSequences }),
    };

    // Add tools if provided
    if (anthropicTools && anthropicTools.length > 0) {
      body.tools = anthropicTools;
      // Map tool_choice to Anthropic format
      if (options?.tool_choice) {
        body.tool_choice = this.convertToolChoice(options.tool_choice);
      }
    }

    try {
      const { response } = await resilientFetch({
        url: `${this.baseUrl}/v1/messages`,
        init: {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'x-api-key': this.apiKey,
            'anthropic-version': '2023-06-01',
            'anthropic-beta': 'prompt-caching-2024-07-31',
          },
          body: JSON.stringify(body),
        },
        providerName: this.name,
        networkConfig: this.networkConfig,
        onRetry: (attempt, delay, error) => {
          console.warn(`[Anthropic] Retry attempt ${attempt} after ${delay}ms: ${error.message}`);
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw this.handleError(response.status, error);
      }

      const data = await response.json() as {
        id: string;
        content: AnthropicContentBlock[];
        stop_reason: string;
        usage: {
          input_tokens: number;
          output_tokens: number;
          cache_creation_input_tokens?: number;
          cache_read_input_tokens?: number;
        };
      };

      // Extract text content and tool calls from content blocks
      const textContent = data.content
        .filter((block): block is { type: 'text'; text: string } => block.type === 'text')
        .map(block => block.text)
        .join('');

      const toolUseBlocks = data.content.filter(
        (block): block is { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> } =>
          block.type === 'tool_use'
      );

      // Convert Anthropic tool_use blocks to our ToolCallResponse format
      const toolCalls: ToolCallResponse[] | undefined = toolUseBlocks.length > 0
        ? toolUseBlocks.map(block => ({
            id: block.id,
            type: 'function' as const,
            function: {
              name: block.name,
              // Anthropic returns input as object, we need to stringify it
              arguments: JSON.stringify(block.input),
            },
          }))
        : undefined;

      return {
        content: textContent,
        stopReason: this.mapStopReason(data.stop_reason),
        usage: {
          inputTokens: data.usage.input_tokens,
          outputTokens: data.usage.output_tokens,
          cacheReadTokens: data.usage.cache_read_input_tokens,
          cacheWriteTokens: data.usage.cache_creation_input_tokens,
        },
        toolCalls,
      };
    } catch (error) {
      if (error instanceof ProviderError) throw error;
      throw new ProviderError(
        `Anthropic request failed: ${(error as Error).message}`,
        this.name,
        'NETWORK_ERROR',
        error as Error
      );
    }
  }

  /**
   * Convert messages to Anthropic's expected format.
   * Handles tool result messages specially.
   */
  private convertMessagesToAnthropicFormat(
    messages: (Message | MessageWithContent)[]
  ): Array<{ role: string; content: string | AnthropicContentBlock[] }> {
    const result: Array<{ role: string; content: string | AnthropicContentBlock[] }> = [];

    for (const msg of messages) {
      // Handle tool result messages
      if ('tool_call_id' in msg && msg.role === 'tool') {
        // Anthropic expects tool results as user messages with tool_result content blocks
        const toolResultBlock: AnthropicContentBlock = {
          type: 'tool_result',
          tool_use_id: msg.tool_call_id!,
          content: typeof msg.content === 'string' ? msg.content : msg.content.map(c => c.text).join(''),
        };

        // Check if the last message is already a user message we can add to
        const lastMsg = result[result.length - 1];
        if (lastMsg && lastMsg.role === 'user' && Array.isArray(lastMsg.content)) {
          lastMsg.content.push(toolResultBlock);
        } else {
          result.push({
            role: 'user',
            content: [toolResultBlock],
          });
        }
        continue;
      }

      // Handle assistant messages with tool calls
      if ('tool_calls' in msg && msg.tool_calls && msg.tool_calls.length > 0) {
        const contentBlocks: AnthropicContentBlock[] = [];

        // Add text content if present
        const textContent = typeof msg.content === 'string'
          ? msg.content
          : msg.content?.map(c => c.text).join('') || '';
        if (textContent) {
          contentBlocks.push({ type: 'text', text: textContent });
        }

        // Convert tool calls to tool_use blocks
        for (const tc of msg.tool_calls) {
          contentBlocks.push({
            type: 'tool_use',
            id: tc.id,
            name: tc.function.name,
            input: JSON.parse(tc.function.arguments),
          });
        }

        result.push({
          role: 'assistant',
          content: contentBlocks,
        });
        continue;
      }

      // Regular user/assistant message
      const content = typeof msg.content === 'string'
        ? msg.content
        : msg.content?.map(c => c.text).join('') || '';

      result.push({
        role: msg.role as 'user' | 'assistant',
        content,
      });
    }

    return result;
  }

  /**
   * Convert OpenAI-format tool definition to Anthropic format.
   */
  private convertToolToAnthropicFormat(tool: ToolDefinitionSchema): AnthropicTool {
    return {
      name: tool.function.name,
      description: tool.function.description,
      input_schema: tool.function.parameters,
    };
  }

  /**
   * Convert tool_choice from OpenAI format to Anthropic format.
   */
  private convertToolChoice(
    choice: ChatOptionsWithTools['tool_choice']
  ): { type: 'auto' | 'any' | 'tool'; name?: string } {
    if (choice === 'auto') {
      return { type: 'auto' };
    }
    if (choice === 'required') {
      return { type: 'any' }; // Anthropic's equivalent of "must use a tool"
    }
    if (choice === 'none') {
      return { type: 'auto' }; // No direct equivalent, use auto
    }
    if (typeof choice === 'object' && choice.function?.name) {
      return { type: 'tool', name: choice.function.name };
    }
    return { type: 'auto' };
  }

  private handleError(status: number, body: string): ProviderError {
    let code: ProviderError['code'] = 'UNKNOWN';
    
    if (status === 401) code = 'AUTHENTICATION_FAILED';
    else if (status === 429) code = 'RATE_LIMITED';
    else if (status === 400) code = 'INVALID_REQUEST';
    else if (status >= 500) code = 'SERVER_ERROR';

    return new ProviderError(
      `Anthropic API error (${status}): ${body}`,
      this.name,
      code
    );
  }

  private mapStopReason(reason: string): ChatResponse['stopReason'] {
    switch (reason) {
      case 'end_turn': return 'end_turn';
      case 'max_tokens': return 'max_tokens';
      case 'stop_sequence': return 'stop_sequence';
      default: return 'end_turn';
    }
  }
}

// =============================================================================
// REGISTRATION
// =============================================================================

registerProvider('anthropic', {
  priority: 1,
  detect: () => hasEnv('ANTHROPIC_API_KEY'),
  create: async () => new AnthropicProvider(),
});

File: attocode/attocode_py/src/attocode/providers/anthropic.py
"""Anthropic API provider using httpx."""

from __future__ import annotations

import json
import os
from collections.abc import AsyncIterator
from typing import Any

import httpx

from attocode.errors import ProviderError
from attocode.types.messages import (
    ChatOptions,
    ChatResponse,
    ContentBlock,
    ImageContentBlock,
    Message,
    MessageWithStructuredContent,
    Role,
    StopReason,
    StreamChunk,
    TextContentBlock,
    TokenUsage,
    ToolCall,
    ToolDefinition,
)

DEFAULT_API_URL = "https://api.anthropic.com/v1/messages"
DEFAULT_MODEL = "claude-sonnet-4-20250514"
DEFAULT_MAX_TOKENS = 8192
API_VERSION = "2023-06-01"

class AnthropicProvider:
    """Anthropic API provider using httpx."""

    def __init__(
        self,
        api_key: str | None = None,
        model: str = DEFAULT_MODEL,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        api_url: str = DEFAULT_API_URL,
        timeout: float = 120.0,
        *,
        extra_headers: dict[str, str] | None = None,
    ) -> None:
        self._api_key = api_key or os.environ.get("ANTHROPIC_API_KEY", "")
        if not self._api_key:
            raise ProviderError("ANTHROPIC_API_KEY not set", provider="anthropic", retryable=False)
        self._model = model
        self._max_tokens = max_tokens
        self._api_url = api_url
        self._timeout = timeout
        self._extra_headers = extra_headers or {}
        self._client = self._create_client()

    def _create_client(self) -> httpx.AsyncClient:
        """Create a fresh httpx client."""
        return httpx.AsyncClient(
            timeout=httpx.Timeout(self._timeout),
            headers={
                "x-api-key": self._api_key,
                "anthropic-version": API_VERSION,
                "content-type": "application/json",
                **self._extra_headers,
            },
        )

    def _ensure_client(self) -> httpx.AsyncClient:
        """Return the client, recreating it if closed."""
        if self._client.is_closed:
            self._client = self._create_client()
        return self._client

    @property
    def name(self) -> str:
        return "anthropic"

    async def chat(
        self,
        messages: list[Message | MessageWithStructuredContent],
        options: ChatOptions | None = None,
    ) -> ChatResponse:
        client = self._ensure_client()
        model = (options and options.model) or self._model
        max_tokens = (options and options.max_tokens) or self._max_tokens

        body: dict[str, Any] = {
            "model": model,
            "max_tokens": max_tokens,
            "messages": self._format_messages(messages),
        }

        if options and options.temperature is not None:
            body["temperature"] = options.temperature

        system_msgs = [m for m in messages if m.role == Role.SYSTEM]
        if system_msgs:
            body["system"] = self._format_system(system_msgs)
            body["messages"] = [m for m in body["messages"] if m.get("role") != "system"]

        if options and options.tools:
            body["tools"] = [self._format_tool(t) for t in options.tools]

        try:
            response = await client.post(self._api_url, json=body)
            response.raise_for_status()
            return self._parse_response(response.json(), model)
        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            raise ProviderError(
                f"Anthropic API error {status}: {e.response.text[:500]}",
                provider="anthropic",
                status_code=status,
                retryable=status in (429, 500, 502, 503, 529),
            ) from e
        except httpx.TimeoutException as e:
            raise ProviderError("Anthropic API timeout", provider="anthropic", retryable=True) from e
        except httpx.RequestError as e:
            raise ProviderError(f"Anthropic request error: {e}", provider="anthropic", retryable=True) from e

    async def chat_stream(
        self,
        messages: list[Message | MessageWithStructuredContent],
        options: ChatOptions | None = None,
    ) -> AsyncIterator[StreamChunk]:
        """Stream a chat response from the Anthropic API."""
        from attocode.integrations.streaming.handler import adapt_anthropic_stream

        client = self._ensure_client()
        model = (options and options.model) or self._model
        max_tokens = (options and options.max_tokens) or self._max_tokens

        body: dict[str, Any] = {
            "model": model,
            "max_tokens": max_tokens,
            "messages": self._format_messages(messages),
            "stream": True,
        }

        if options and options.temperature is not None:
            body["temperature"] = options.temperature

        system_msgs = [m for m in messages if m.role == Role.SYSTEM]
        if system_msgs:
            body["system"] = self._format_system(system_msgs)
            body["messages"] = [m for m in body["messages"] if m.get("role") != "system"]

        if options and options.tools:
            body["tools"] = [self._format_tool(t) for t in options.tools]

        try:
            async with client.stream("POST", self._api_url, json=body) as response:
                if response.status_code >= 400:
                    # Must read body INSIDE async-with before response closes
                    await response.aread()
                    status = response.status_code
                    error_body = response.text[:500]
                    raise ProviderError(
                        f"Anthropic API error {status}: {error_body}",
                        provider="anthropic",
                        status_code=status,
                        retryable=status in (429, 500, 502, 503, 529),
                    )
                async for chunk in adapt_anthropic_stream(response.aiter_lines()):
                    yield chunk
        except ProviderError:
            raise
        except httpx.TimeoutException as e:
            raise ProviderError("Anthropic API timeout", provider="anthropic", retryable=True) from e
        except httpx.RequestError as e:
            raise ProviderError(f"Anthropic request error: {e}", provider="anthropic", retryable=True) from e

    def _format_messages(self, messages: list[Message | MessageWithStructuredContent]) -> list[dict[str, Any]]:
        result: list[dict[str, Any]] = []
        for msg in messages:
            if msg.role == Role.SYSTEM:
                continue
            formatted = self._format_single(msg)
            if formatted:
                result.append(formatted)
        return result

    def _format_single(self, msg: Message | MessageWithStructuredContent) -> dict[str, Any] | None:
        if msg.role == Role.TOOL:
            return {
                "role": "user",
                "content": [{"type": "tool_result", "tool_use_id": msg.tool_call_id, "content": str(msg.content)}],
            }
        if msg.role == Role.ASSISTANT and msg.tool_calls:
            blocks: list[dict[str, Any]] = []
            content = msg.content
            if isinstance(content, str) and content:
                blocks.append({"type": "text", "text": content})
            for tc in msg.tool_calls:
                blocks.append({"type": "tool_use", "id": tc.id, "name": tc.name, "input": tc.arguments})
            return {"role": "assistant", "content": blocks}
        content = msg.content
        if isinstance(content, list):
            blocks = []
            for block in content:
                if isinstance(block, TextContentBlock):
                    b: dict[str, Any] = {"type": "text", "text": block.text}
                    if block.cache_control:
                        b["cache_control"] = {"type": block.cache_control.type}
                    blocks.append(b)
                elif isinstance(block, ImageContentBlock):
                    blocks.append({
                        "type": "image",
                        "source": {"type": block.source.type, "media_type": block.source.media_type, "data": block.source.data},
                    })
            return {"role": str(msg.role), "content": blocks}
        return {"role": str(msg.role), "content": content}

    def _format_system(self, msgs: list[Message | MessageWithStructuredContent]) -> str | list[dict[str, Any]]:
        if len(msgs) == 1 and isinstance(msgs[0].content, str):
            return msgs[0].content
        blocks: list[dict[str, Any]] = []
        for msg in msgs:
            if isinstance(msg.content, str):
                blocks.append({"type": "text", "text": msg.content})
        return blocks

    def _format_tool(self, tool: ToolDefinition) -> dict[str, Any]:
        return {"name": tool.name, "description": tool.description, "input_schema": tool.parameters}

    def _parse_response(self, data: dict[str, Any], model: str) -> ChatResponse:
        text_parts: list[str] = []
        tool_calls: list[ToolCall] = []
        thinking: str | None = None

        for block in data.get("content", []):
            if block["type"] == "text":
                text_parts.append(block["text"])
            elif block["type"] == "tool_use":
                tool_calls.append(ToolCall(id=block["id"], name=block["name"], arguments=block.get("input", {})))
            elif block["type"] == "thinking":
                thinking = block.get("thinking", "")

        usage_data = data.get("usage", {})
        usage = TokenUsage(
            input_tokens=usage_data.get("input_tokens", 0),
            output_tokens=usage_data.get("output_tokens", 0),
            total_tokens=usage_data.get("input_tokens", 0) + usage_data.get("output_tokens", 0),
            cache_read_tokens=usage_data.get("cache_read_input_tokens", 0),
            cache_creation_tokens=usage_data.get("cache_creation_input_tokens", 0),
        )
        from attocode.providers.base import get_model_pricing

        pricing = get_model_pricing(model)
        usage.cost = pricing.estimate_cost(
            usage.input_tokens, usage.output_tokens, usage.cache_read_tokens,
        )

        stop = data.get("stop_reason", "end_turn")
        stop_reason = StopReason.TOOL_USE if stop == "tool_use" else (StopReason.MAX_TOKENS if stop == "max_tokens" else StopReason.END_TURN)

        return ChatResponse(
            content="\n".join(text_parts),
            tool_calls=tool_calls or None,
            usage=usage,
            model=model,
            stop_reason=stop_reason,
            thinking=thinking,
        )

    async def close(self) -> None:
        await self._client.aclose()
